imports/wasi_snapshot_preview1/wrapper.go

```diff
@@ -0,0 +1,234 @@
+	}
+
+	// outParams are i32, even if the corresponding outParamLengths are not,
+	// because memory offsets are i32.
+	outParamTypes := make([]api.ValueType, outLen)
+	for i := 0; i < outLen; i++ {
+		outParamTypes[i] = i32
+	}
+	outParamIdx := len(goFunc.ParamTypes)
+
+	proxy := &wasm.HostFunc{
+		ExportNames: []string{exportName},
+		Name:        exportName,
+		ParamTypes:  append(goFunc.ParamTypes, outParamTypes...),
+		ParamNames:  append(goFunc.ParamNames, outParamNames...),
+		ResultTypes: []api.ValueType{i32},
+		ResultNames: []string{"errno"},
+		Code:        &wasm.Code{IsHostFunction: true, LocalTypes: []api.ValueType{i32}},
+	}
+
+	// Determine the index of a temporary local used for i32 manipulation. This
+	// will be directly after the parameters.
+	localValIdxI32 := byte(len(proxy.ParamTypes))
+
+	// Only allocate an i64 local if an out type is i64
+	localI64Temp := localValIdxI32 + 1
+	for _, t := range outTypes {
+		if t == i64 {
+			proxy.Code.LocalTypes = append(proxy.Code.LocalTypes, i64)
+			break
+		}
+	}
+
+	// Get the current memory size in bytes, to validate parameters.
+	body := memBytes                                         // stack: [memSizeBytes]
+	body = append(body, wasm.OpcodeLocalTee, localValIdxI32) // stack: [memSizeBytes]
+
+	// Write code to verify out parameters are in range to write values.
+	// Notably, this validates prior to calling the host function.
+	body = append(body, compileMustValidMemOffset(outParamIdx, outTypes[0])...)
+	for i := 1; i < outLen; i++ {
+		body = append(body, wasm.OpcodeLocalGet, localValIdxI32) // stack: [memSizeBytes]
+		body = append(body, compileMustValidMemOffset(outParamIdx+i, outTypes[i])...)
+	}
+
+	// Now, it is safe to call the go function
+	for i := range goFunc.ParamTypes {
+		body = append(body, wasm.OpcodeLocalGet, byte(i)) // stack: [p1, p2...]
+	}
+	body = append(body, wasm.OpcodeCall, fakeFuncIdx) // stack: [r1, r2... errno]
+	// Capture the position in the wasm ProxyFunc rewrites with a real index.
+	fakeFuncIdxPos := len(body) - 1
+
+	// On failure, return the errno
+	body = append(body, compileMustErrnoSuccess(localValIdxI32)...) // stack: [r1, r2...]
+
+	// On success, results to write to memory are in backwards order.
+	for i := outLen - 1; i >= 0; i-- {
+		outType := outTypes[i]
+		localValIdx := localValIdxI32
+		switch outType {
+		case i32:
+		case i64:
+			localValIdx = localI64Temp
+		default:
+			panic(fmt.Errorf("TODO: unsupported outType: %v", outType))
+		}
+		localOffsetIdx := byte(outParamIdx + i)
+		body = append(body, compileStore(localOffsetIdx, localValIdx, outType)...)
+	}
+
+	// Finally, add the success error code to the stack.
+	body = append(body, wasm.OpcodeI32Const, success) // stack: [success]
+	body = append(body, wasm.OpcodeEnd)
+
+	// Assign the wasm generated as the proxy's body
+	proxy.Code.Body = body
+	return &wasm.ProxyFunc{Proxy: proxy, Proxied: goFunc, CallBodyPos: fakeFuncIdxPos}
```

Comment:
just realized that we haven't checked if the function indexes replaced by ProxyFunc never exceed one-byte leb128 encoding. This PR makes it more likely to happen that we have 2-bytes long function indexes?

Reply:
It isn't enforced here, it is enforced where we make a host module. I don't think we'll run into this in the near term because this is internal and our biggest host module is wasi. We might run into it, if we make this non-internal later, or use it in emscripten which could make massive amounts of things.

from internal/wasm/host.go
```
			proxiedIdx := len(funcNames)
			if proxiedIdx > maxProxiedFuncIdx {
				return errors.New("TODO: proxied funcidx larger than one byte")
			}
```

Reply:
oh cool

Reply:
so concretely, old wasi is 45 functions, and many aren't implemented. Even if we could proxy them all it would reach 90 in that module. 127 is the max. I think yeah it would be emscripten that if we start using this approach would end up hitting it. The result of that would be some work in the host module code to insert bytes for the larger func index. We did something similar for the text format back in the day, so not a big deal to do when we get to it.